File: src/commands/retag.ts
import { readFile, writeFile } from 'fs/promises';
import type { Command } from 'commander';
import { join } from 'path';
import chalk from 'chalk';
import { glob } from 'tinyglobby';
import * as Gherkin from '@cucumber/gherkin';
import * as Messages from '@cucumber/messages';

interface RetagOptions {
  from: string;
  to: string;
  dryRun?: boolean;
  cwd?: string;
}

interface RetagResult {
  success: boolean;
  fileCount: number;
  occurrenceCount: number;
  message?: string;
  files?: string[];
  error?: string;
}

export async function retag(options: RetagOptions): Promise<RetagResult> {
  const { from, to, dryRun = false, cwd = process.cwd() } = options;

  // Validate required parameters
  if (!from || !to) {
    return {
      success: false,
      fileCount: 0,
      occurrenceCount: 0,
      error: 'Both --from and --to are required',
    };
  }

  // Validate tag format for 'to' parameter
  if (!to.startsWith('@')) {
    return {
      success: false,
      fileCount: 0,
      occurrenceCount: 0,
      error: `Invalid tag format: "${to}". Valid format is @lowercase-with-hyphens`,
    };
  }

  if (!/^@[a-z0-9-#]+$/.test(to)) {
    return {
      success: false,
      fileCount: 0,
      occurrenceCount: 0,
      error: `Invalid tag format: "${to}". Valid format is @lowercase-with-hyphens`,
    };
  }

  try {
    // Get all feature files
    const files = await glob(['spec/features/**/*.feature'], {
      cwd,
      absolute: false,
    });

    if (files.length === 0) {
      return {
        success: true,
        fileCount: 0,
        occurrenceCount: 0,
        message: 'No feature files found',
      };
    }

    // Find all files containing the 'from' tag
    const matchingFiles: Array<{ file: string; occurrences: number }> = [];

    for (const file of files) {
      const filePath = join(cwd, file);
      const content = await readFile(filePath, 'utf-8');

      // Simple regex to count tag occurrences
      // Match tag as whole word (preceded by whitespace or start, followed by whitespace or end)
      const escapedFrom = from.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
      const tagRegex = new RegExp(`(^|\\s)${escapedFrom}(?=\\s|$)`, 'gm');
      const matches = content.match(tagRegex);

      if (matches && matches.length > 0) {
        matchingFiles.push({
          file,
          occurrences: matches.length,
        });
      }
    }

    // No files found with the tag
    if (matchingFiles.length === 0) {
      return {
        success: false,
        fileCount: 0,
        occurrenceCount: 0,
        error: `Tag ${from} not found in any feature files`,
      };
    }

    const totalOccurrences = matchingFiles.reduce(
      (sum, f) => sum + f.occurrences,
      0
    );

    // Dry run - just report what would happen
    if (dryRun) {
      return {
        success: true,
        fileCount: matchingFiles.length,
        occurrenceCount: totalOccurrences,
        message: `Would rename ${from} to ${to} in ${matchingFiles.length} file(s) (${totalOccurrences} occurrence(s))`,
        files: matchingFiles.map(f => f.file),
      };
    }

    // Perform renaming
    for (const { file } of matchingFiles) {
      const filePath = join(cwd, file);
      let content = await readFile(filePath, 'utf-8');

      // Replace all occurrences of the tag
      // Match tag as whole word (preceded by whitespace or start, followed by whitespace or end)
      const escapedFrom = from.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
      const tagRegex = new RegExp(`(^|\\s)${escapedFrom}(?=\\s|$)`, 'gm');
      content = content.replace(tagRegex, `$1${to}`);

      // Validate the new content is valid Gherkin
      const builder = new Gherkin.AstBuilder(Messages.IdGenerator.uuid());
      const matcher = new Gherkin.GherkinClassicTokenMatcher();
      const parser = new Gherkin.Parser(builder, matcher);

      try {
        parser.parse(content);
      } catch (error: any) {
        return {
          success: false,
          fileCount: 0,
          occurrenceCount: 0,
          error: `Validation failed after renaming in ${file}: ${error.message}`,
        };
      }

      // Write the modified content
      await writeFile(filePath, content, 'utf-8');
    }

    return {
      success: true,
      fileCount: matchingFiles.length,
      occurrenceCount: totalOccurrences,
      message: `Renamed ${from} to ${to} in ${matchingFiles.length} file(s) (${totalOccurrences} occurrence(s)). All modified files validated successfully.`,
      files: matchingFiles.map(f => f.file),
    };
  } catch (error: any) {
    return {
      success: false,
      fileCount: 0,
      occurrenceCount: 0,
      error: error.message,
    };
  }
}

export async function retagCommand(options: {
  from?: string;
  to?: string;
  dryRun?: boolean;
}): Promise<void> {
  try {
    const result = await retag({
      from: options.from || '',
      to: options.to || '',
      dryRun: options.dryRun,
    });

    if (!result.success) {
      console.error(chalk.red('Error:'), result.error);
      process.exit(1);
    }

    if (options.dryRun && result.files) {
      console.log(chalk.yellow('Dry run mode - no files modified'));
      console.log(
        chalk.cyan(
          `\nWould rename ${options.from} to ${options.to} in ${result.fileCount} file(s) (${result.occurrenceCount} occurrence(s)):\n`
        )
      );

      for (const file of result.files) {
        console.log(chalk.gray(`  - ${file}`));
      }
    } else if (result.files && result.files.length > 0) {
      console.log(chalk.green(`✓ ${result.message}`));
      console.log(chalk.gray('\nModified files:'));
      for (const file of result.files) {
        console.log(chalk.gray(`  - ${file}`));
      }
    } else {
      console.log(chalk.yellow(result.message));
    }

    process.exit(0);
  } catch (error: any) {
    console.error(chalk.red('Error:'), error.message);
    process.exit(1);
  }
}

export function registerRetagCommand(program: Command): void {
  program
    .command('retag')
    .description('Bulk rename tags across all feature files')
    .option('--from <tag>', 'Tag to rename from (e.g., @old-tag)')
    .option('--to <tag>', 'Tag to rename to (e.g., @new-tag)')
    .option('--dry-run', 'Preview changes without making modifications')
    .action(retagCommand);
}

File: rust/fspec-core/src/commands/retag.rs
//! `retag` — Rust port of `src/commands/retag.ts` (RPC-293).
//!
//! Bulk-renames a tag (`--from` → `--to`) across every
//! `spec/features/**/*.feature` file. Matching is a whole-token text replace:
//! the `from` tag must be preceded by start-of-line or whitespace and followed
//! by whitespace or end-of-line (parity with the TS
//! `new RegExp('(^|\\s)' + escapedFrom + '(?=\\s|$)', 'gm')`). The preceding
//! whitespace is preserved and only the tag bytes are rewritten (TS `$1${to}`).
//!
//! The recursive `spec/features` walk reuses
//! [`crate::io::feature_glob::glob_feature_files`]; a `DirectoryNotFound`
//! result is mapped to an empty list so the canonical TS `"No feature files
//! found"` message is preserved when `spec/features/` is absent. After each
//! rename the new content is re-parsed with
//! [`crate::io::gherkin::parse_feature_lenient`] (parity with the TS Gherkin
//! re-parse guard) before the file is written.
//!
//! ## Result envelope
//! Like `delete_features.rs` (RPC-218), retag.ts RETURNS a `RetagResult`
//! object (never throws), so EVERY business outcome — including the
//! validation rejections and the not-found case — is returned as the inner
//! JSON envelope `{success, fileCount, occurrenceCount, message?, files?,
//! error?}` via `Ok(json)`. The dispatcher envelope therefore succeeds; the
//! real outcome is carried by the inner `success`/`error` fields, and the CLI
//! bridge owns every rendering decision.
//!
//! ## Two-front-doors
//! Both the LLM-facing dispatcher AND the standalone Rust binary's clap
//! subcommand call this single function. The CLI bridge at
//! `rust/fspec/src/retag.rs` is JSON marshalling + rendering only.

use std::path::Path;

use serde::Deserialize;
use serde_json::{json, Value};

use crate::error::FspecCoreError;
use crate::io::feature_glob::glob_feature_files;
use crate::io::gherkin::parse_feature_lenient;

#[derive(Debug, Default, Deserialize)]
#[serde(default, rename_all = "camelCase")]
struct RetagArgs {
    from: Option<String>,
    to: Option<String>,
    dry_run: bool,
}

/// Dispatcher entry point.
pub async fn run(args_json: &str, project_root: &Path) -> Result<String, FspecCoreError> {
    let args: RetagArgs =
        serde_json::from_str(args_json).map_err(|e| FspecCoreError::InvalidArgs {
            command: "retag",
            reason: format!("failed to parse args: {e}"),
        })?;

    let from = args.from.unwrap_or_default();
    let to = args.to.unwrap_or_default();

    // ---- Require both --from and --to (TS parity, lines 30-37) ----
    // TS treats empty strings as falsy via `!from || !to`.
    if from.is_empty() || to.is_empty() {
        return ok(json!({
            "success": false,
            "fileCount": 0,
            "occurrenceCount": 0,
            "error": "Both --from and --to are required",
        }));
    }

    // ---- Validate target tag format (TS parity, lines 40-56) ----
    if !to.starts_with('@') || !valid_to_tag(&to) {
        return ok(json!({
            "success": false,
            "fileCount": 0,
            "occurrenceCount": 0,
            "error": format!(
                "Invalid tag format: \"{to}\". Valid format is @lowercase-with-hyphens"
            ),
        }));
    }

    // ---- Enumerate feature files (DirectoryNotFound → empty list) ----
    let files = match glob_feature_files(project_root) {
        Ok(f) => f,
        Err(FspecCoreError::DirectoryNotFound { .. }) => Vec::new(),
        Err(other) => return Err(other),
    };

    if files.is_empty() {
        return ok(json!({
            "success": true,
            "fileCount": 0,
            "occurrenceCount": 0,
            "message": "No feature files found",
        }));
    }

    // ---- Find all files containing the 'from' tag ----
    let mut matching: Vec<(String, usize)> = Vec::new();
    for file in &files {
        let abs = project_root.join(file);
        let content = match std::fs::read_to_string(&abs) {
            Ok(s) => s,
            Err(source) => {
                return Err(FspecCoreError::Io {
                    command: "retag",
                    source,
                })
            }
        };
        let occurrences = count_tag_matches(&content, &from);
        if occurrences > 0 {
            matching.push((file.clone(), occurrences));
        }
    }

    // ---- No files contain the tag ----
    if matching.is_empty() {
        return ok(json!({
            "success": false,
            "fileCount": 0,
            "occurrenceCount": 0,
            "error": format!("Tag {from} not found in any feature files"),
        }));
    }

    let total_occurrences: usize = matching.iter().map(|(_, n)| n).sum();
    let file_count = matching.len();
    let matched_files: Vec<String> = matching.iter().map(|(f, _)| f.clone()).collect();

    // ---- Dry run: report without modifying ----
    if args.dry_run {
        return ok(json!({
            "success": true,
            "fileCount": file_count,
            "occurrenceCount": total_occurrences,
            "message": format!(
                "Would rename {from} to {to} in {file_count} file(s) ({total_occurrences} occurrence(s))"
            ),
            "files": matched_files,
        }));
    }

    // ---- Perform renaming ----
    for (file, _) in &matching {
        let abs = project_root.join(file);
        let content = std::fs::read_to_string(&abs).map_err(|source| FspecCoreError::Io {
            command: "retag",
            source,
        })?;
        let new_content = replace_tags(&content, &from, &to);

        // Validate the new content is still parseable Gherkin (TS re-parse
        // guard, lines 132-146).
        if let Err(e) = parse_feature_lenient(&new_content) {
            return ok(json!({
                "success": false,
                "fileCount": 0,
                "occurrenceCount": 0,
                "error": format!("Validation failed after renaming in {file}: {e}"),
            }));
        }

        std::fs::write(&abs, new_content.as_bytes()).map_err(|source| FspecCoreError::Io {
            command: "retag",
            source,
        })?;
    }

    ok(json!({
        "success": true,
        "fileCount": file_count,
        "occurrenceCount": total_occurrences,
        "message": format!(
            "Renamed {from} to {to} in {file_count} file(s) ({total_occurrences} occurrence(s)). \
             All modified files validated successfully."
        ),
        "files": matched_files,
    }))
}

/// Serialise an inner-envelope value to the `Ok(String)` returned by the
/// dispatcher entry point.
fn ok(value: Value) -> Result<String, FspecCoreError> {
    serde_json::to_string(&value).map_err(|e| FspecCoreError::InvalidArgs {
        command: "retag",
        reason: format!("failed to serialise response: {e}"),
    })
}

/// `^@[a-z0-9-#]+$` — the TS target-tag format check.
fn valid_to_tag(to: &str) -> bool {
    let rest = match to.strip_prefix('@') {
        Some(r) => r,
        None => return false,
    };
    !rest.is_empty()
        && rest
            .bytes()
            .all(|b| b.is_ascii_lowercase() || b.is_ascii_digit() || b == b'-' || b == b'#')
}

/// Byte-start indices of every whole-token occurrence of `from` in `content`,
/// reproducing the `(^|\s)<from>(?=\s|$)` global multiline regex: each match
/// is preceded by start-of-string or a whitespace char and followed by a
/// whitespace char or end-of-string. The regex restarts at the END of the tag
/// (the trailing-boundary lookahead is non-consuming), so a whitespace char
/// can simultaneously be one match's lookahead and the next match's leading
/// boundary.
fn tag_match_starts(content: &str, from: &str) -> Vec<usize> {
    let mut out: Vec<usize> = Vec::new();
    if from.is_empty() {
        return out;
    }
    let flen = from.len();
    let mut i = 0;
    while let Some(rel) = content[i..].find(from) {
        let start = i + rel;
        let end = start + flen;
        let prev_ok = start == 0
            || content[..start]
                .chars()
                .next_back()
                .is_some_and(char::is_whitespace);
        let next_ok = end == content.len()
            || content[end..]
                .chars()
                .next()
                .is_some_and(char::is_whitespace);
        if prev_ok && next_ok {
            out.push(start);
            i = end;
        } else {
            // Advance one char past `start` to keep scanning.
            let step = content[start..].chars().next().map_or(1, char::len_utf8);
            i = start + step;
        }
    }
    out
}

/// Count whole-token occurrences of `from`.
fn count_tag_matches(content: &str, from: &str) -> usize {
    tag_match_starts(content, from).len()
}

/// Replace every whole-token occurrence of `from` with `to`, preserving all
/// surrounding bytes (parity with the TS `$1${to}` replacement).
fn replace_tags(content: &str, from: &str, to: &str) -> String {
    let starts = tag_match_starts(content, from);
    if starts.is_empty() {
        return content.to_string();
    }
    let flen = from.len();
    let mut result = String::with_capacity(content.len());
    let mut last = 0;
    for &start in &starts {
        result.push_str(&content[last..start]);
        result.push_str(to);
        last = start + flen;
    }
    result.push_str(&content[last..]);
    result
}

#[cfg(test)]
mod tests {
    #![allow(clippy::unwrap_used, clippy::expect_used, clippy::panic)]
    use super::*;

    #[test]
    fn args_parse_camel_case() {
        let a: RetagArgs =
            serde_json::from_str(r#"{"from":"@a","to":"@b","dryRun":true}"#).unwrap();
        assert_eq!(a.from.as_deref(), Some("@a"));
        assert_eq!(a.to.as_deref(), Some("@b"));
        assert!(a.dry_run);
    }

    #[test]
    fn valid_to_tag_accepts_and_rejects() {
        assert!(valid_to_tag("@in-progress"));
        assert!(valid_to_tag("@phase-1"));
        assert!(valid_to_tag("@a#b"));
        assert!(!valid_to_tag("WIP"));
        assert!(!valid_to_tag("@"));
        assert!(!valid_to_tag("@UPPER"));
    }

    #[test]
    fn counts_whole_token_only() {
        // "@wip" appears as a token twice, plus "@wipe" should NOT match.
        let content = "@wip\nFeature: X\n  @wip\n  @wipe matters not\n";
        assert_eq!(count_tag_matches(content, "@wip"), 2);
    }

    #[test]
    fn counts_adjacent_tags_separated_by_single_space() {
        assert_eq!(count_tag_matches("@a @a @a", "@a"), 3);
    }

    #[test]
    fn replace_preserves_surrounding_whitespace() {
        let content = "  @wip @smoke\n";
        assert_eq!(
            replace_tags(content, "@wip", "@in-progress"),
            "  @in-progress @smoke\n"
        );
    }

    #[test]
    fn replace_does_not_touch_substring_matches() {
        let content = "@wipe @wip\n";
        assert_eq!(replace_tags(content, "@wip", "@x"), "@wipe @x\n");
    }
}
